plugin/pulpcore/plugin/changeset/iterator.py

```diff
@@ -166,13 +169,21 @@ def _iter(self):
         Yields:
             Download: The appropriate download object.
         """
-        for artifact in ArtifactIterator(self.content):
-            if self.deferred or artifact.model.downloaded:
+        for remote_artifact in ArtifactIterator(self.content):
+            q = Q()
+            for algorithm in hashlib.algorithms_guaranteed:
+                if getattr(remote_artifact.model, algorithm):
+                    kwargs = {algorithm: getattr(remote_artifact.model, algorithm)}
+                    q |= Q(**kwargs)
+            try:
+                artifact = Artifact.objects.get(q)
```

Comment:
This query needs to be done in batches.  For performance, we cannot be hitting the DB for every artifact.  Other code in the _changeset_ package does this.